### PR TITLE
[3.3][DFSM] Add DeletionPolicy to EfsSettings and FsxLustreSettings

### PR DESCRIPTION
**Pull request by gmarciani**
### Description of changes
Add `DeletionPolicy=Retain|Delete` to `EfsSettings` and `FsxLustreSettings`.
Not added to `EbsSettings` because already there.
Not added to `FsxOntapSettings` and `FsxOpenZfsSettings` because in these cases only external file systems are supported so `DeletionPolicy` should not be there.

Added unit tests covering shared storage resources within the generated template. With this PR I focused such unit tests on deletion policy and security groups, but they can be easily extended to cover other properties.

### Tests
1. Unit tests
1. Integration tests: 15 out of 20 storage tests succeeded but the following: 
    1. test_retain_on_deletion: the failure is not related to this change but to the fact that we moved resource names from positional ids to storage names, .e.g "EBS0" does not exist anymore within the template.
   2. test_ebs_existing
   3. test_multiple_efs
   4. test_multiple_fsx

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

Signed-off-by: Giacomo Marciani <mgiacomo@amazon.com>

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,9 @@ CHANGELOG
 3.x.x
 -----
 
+**ENHANCEMENTS**
+- Add new configuration parameter `SharedStorage/[Efs,FsxLustre]Settings/DeletionPolicy` to support storage retention on deletion.
```
Comment (enrico-usai) [inline CHANGELOG.md]: I'd write: 
Add `DeletionPolicy` configuration parameter for EFS and FSx for Lustre shared storage.

File: cli/tests/pcluster/templates/test_shared_storage.py [hunk 2]
```diff
@@ -20,15 +21,112 @@
...
+        generated_template, type="AWS::EFS::MountTarget", properties={"FileSystemId": {"Ref": file_system_name}}
+    )
+
+    assert_that(mount_targets).is_length(1)
+
+    mount_target = next(iter(mount_targets.values()))
+
+    assert_that(mount_target["DeletionPolicy"]).is_equal_to(deletion_policy)
+
+    mount_target_sg_name = mount_target["Properties"]["SecurityGroups"][0]["Ref"]
```
Comment (enrico-usai) [inline cli/tests/pcluster/templates/test_shared_storage.py:90]: nit: I'd keep the actions group together in a logical way rather than having all of them separated by a newline.
I mean:

```python
 mount_target = next(iter(mount_targets.values()))
 assert_that(mount_target["DeletionPolicy"]).is_equal_to(deletion_policy)

mount_target_sg_name = mount_target["Properties"]["SecurityGroups"][0]["Ref"]
mount_target_sg = generated_template["Resources"][mount_target_sg_name]
assert_that(mount_target_sg["DeletionPolicy"]).is_equal_to(deletion_policy)
```

in the first group of code we're getting the info for the first assert, in the second group we're getting the info for the second assert, etc

**Review by enrico-usai — DISMISSED** [review]
Thanks for adding this new set of unit tests.
The patch is fine by me, but there are some unit tests failing.

**Comment by gmarciani** [comment]
> Thanks for adding this new set of unit tests. The patch is fine by me, but there are some unit tests failing.

Thanks, fixed.

File: CHANGELOG.md [hunk 3]
```diff
@@ -4,6 +4,10 @@ CHANGELOG
 3.x.x
 -----
 
+**ENHANCEMENTS**
+- Add new configuration parameter `DeletionPolicy` for EFs and FSx for Lustre shared storage
```
Comment (enrico-usai) [inline CHANGELOG.md:8]: EFs --> EFS

**Review by enrico-usai — APPROVED** [review]

File: CHANGELOG.md (hunk 3 above)
Comment (chenwany) [reply CHANGELOG.md:8]: I am going to merge this PR first and address this change in my PR

File: cli/src/pcluster/templates/cluster_stack.py [hunk 4]
```diff
@@ -654,10 +698,12 @@ def _add_fsx_storage(self, id: str, shared_fsx: BaseSharedFsx):
                 file_system_type=LUSTRE,
                 storage_type=shared_fsx.fsx_storage_type,
                 subnet_ids=self.config.compute_subnet_ids,
-                security_group_ids=self._get_compute_security_groups(),
+                security_group_ids=[sg.ref for sg in file_system_security_groups],
```
Comment (hanwen-pcluste) [inline cli/src/pcluster/templates/cluster_stack.py:701]: Why this change was made? After the change, if a user would avoid ParallelCluster creating any security groups, besides setting `SecurityGroups` in `Queues`, they have to avoid using any managed EFS or FSx.
